---
title: "Analysis of BLAST Results"
author: "Katelyn B. Kukar"
date: "October 6, 2017"
output: github_document
---

# Introduction


Creating alternative ways of identification is highly important within the field of Forensics because this would allow for more specific analysis of potential suspects. By being able to compare blood, fingerprints, and other means of testing methods allows for a higher level of statistical significance for true suspect identification. Creating alternative methods is also important since some investigations of suspects of interest lack both DNA evidence and individual fingerprints allowing for ambiguity that could be resolved via other testing means. In Fierer et al., scientists explain that bacteria colonized on individual hands represent highly specific and diverse communities. They believe that "because skin bacterial communities are personalized, [they] could use the residual skin bacteria left on objects for forensic identification, matching the bacteria on the object to the skin-associated bacteria of the individual who touched the object," (Fierer et al.). 

Within this study we analyze data collected from the Fierer Lab in Colorado through sequencing and computational techniques. The information gathered from the sequencing sets will be evaluated to further evidence towards the validity of utilizing such methods within forensic analysis. Specifically this analysis will utilize prior knowledge geared towards differences between the male and female sex. We hypothesize the female sex will have signficantlly different data than that of the male sex due to physiological and behavioral differences. Certain assumptions for this hypothesis include males having traditionally different hand shape and size, body hair, and fat percentage which constitue physiological differences. Assumptions of behavior for these differences are based on hygiene patterns experienced by males and females due to cultural biases towards the feminine construct.


# Methods

Sample origin, sequencing and computational techniques were utilized to analyze sequencing sets collected from human participants and coinciding mice which were touched by each participant in the study. 


## Sample origin and sequencing



## Computational

In order to analyze the sequences received we used computational programs to evaluate species of interest. 
  First, we analyzed the Fastq files which stored the biological sequences ran by evaluating the quality control number contained. We looked at the coressponding quality score in a chart to ensure the sequencing data was coherent and applicable to our study (no outliers due to contamination or fault sequencing techniques).
  Second, we took sequences gathered in Fastq files and ran them against a program to trim and align each length for optimal analysis. We then used a secondary program to 

# Results

The results section was primarily used to extrapolate data in regards to female participants and the specific bacterial communities identified. This section outlines statistical analyses used to filter through the data set to identify the specific bacterial communties and where they arose against male counterpots.

```{r load-libraries, message = FALSE, echo=FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the intall.packages() function
# or with the 'Packages' pane in RStudio

# load packages
library("dplyr")
library("tidyr")
library("knitr")
library("ggplot2")
```

```{r make-read-in-data-function, echo=FALSE}
# Output format from BLAST is as detailed on:
# https://www.ncbi.nlm.nih.gov/books/NBK279675/
# In this case, we used: '10 sscinames std'
# 10 means csv format
# sscinames means unique Subject Scientific Name(s), separated by a ';'
# std means the standard set of result columns, which are:
# 'qseqid sseqid pident length mismatch
# gapopen qstart qend sstart send evalue bitscore',


# this function takes as input a quoted path to a BLAST result file
# and produces as output a dataframe with proper column headers
# and the 'qseqid' column split into sample and seq number
read_blast_output <- function(filename) {
  data_in <- read.csv(filename,
                      header = FALSE, # files don't have column names in them
                      col.names = c("sscinames", # unique Subject Sci Name(s)
                                    "qseqid",    # Query Seq-id
                                    "sseqid",    # Subject Seq-id
                                    "pident",    # Percntge of identical matches
                                    "length",    # Alignment length
                                    "mismatch",  # Number of mismatches
                                    "gapopen",   # Number of gap openings
                                    "qstart",    # Start of alignment in query
                                    "qend",      # End of alignment in query
                                    "sstart",    # Start of alignment in subj
                                    "send",      # End of alignment in subject
                                    "evalue",    # Expect value
                                    "bitscore"))  # Bit score

  # Next we want to split the query sequence ID into
  # Sample and Number components so we can group by sample
  # They originally look like "ERR1942280.1"
  # and we want to split that into two columns: "ERR1942280" and "1"
  # we can use the separate() function from the tidyr library to do this
  # Note that we have to double escape the period for this to work
  # the syntax is
  # separate(column_to_separate,
  # c("New_column_name_1", "New_column_name_2"),
  # "seperator")
  data_in <- data_in %>%
    separate(qseqid, c("sample_name", "sample_number"), "\\.")
}
```

```{r read-in-BLAST-data,echo=FALSE}
# this makes a vector of all the BLAST output file names, including
# the name(s) of the directories they are in
files_to_read_in <- list.files(path = "output/blast",
                               full.names = TRUE)

# We need to create an empty matrix with the right number of columns
# so that we can rbind() each dataset on to it
joined_blast_data <- matrix(nrow = 0,
                            ncol = 14)

# now we loop over each of the files in the list and append them
# to the bottom of the 'joined_blast_data' object
# we do this with the rbind() function and the function we
# made earlier to read in the files, read_blast_output()
for (filename in files_to_read_in) {
  joined_blast_data <- rbind(joined_blast_data,
                             read_blast_output(filename))
}
```

```{r read-in-metadata-and-join, echo=FALSE}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the `sep = "\t"` tells the function that the data are tab-delimited
# and the `stringsAsFactors = FALSE` tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                                 "fierer_forensic_hand_mouse_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE)

# Finally we use the left_join() function from dplyr to merge or 'join' the
# combined data and metadata into one big table, so it's easier to work with
# in R the `by = c("Run_s" = "sample_name")` syntax tells R which columns
# to match up when joining the datasets together
joined_blast_data_metadata <- metadata_in %>%
  left_join(joined_blast_data,
            by = c("Run_s" = "sample_name"))
```

###Extrapolating Human Female Data from Meta Data

The data first needed to be subsetted by human genomic data which was done though the extrapolation of "Sebum," utilizing the percent identification for each sequence.

**Figure 1a.** Histogram of Percent Identity of Human Environmental Material within Whole Data

```{r histograms, echo=FALSE}
# Here we're using the dply piping syntax to select a subset of rows matching a
# criteria we specify (using the filter) function, and then pull out a column
# from the data to make a histogram. We don't need to tell the hist() function
# which data to use, because that's piped in, but we do have to give the
# hist() function the title and axis label we'd like to use for the figure

# Histogram to visualize only human samples utilizing Human Skin Oil
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  pull(pident) %>%
  hist(main = "Percent Identity of Human Environmental Material",
       xlab = "Percent Identity",
       col = "darkseagreen2")
```

Once primarily human genomic data was extrapolated, only the female data was further filtered out utilizing the percent indentification for each sequence

**Figure 1b.** Histogram of Percent Identity of Human Environmental Material Analyzed for Females

```{r, echo=FALSE}
# Histogram to visualize sex of human hosts
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "female") %>%
  pull(pident) %>%
  hist(main = "Percent Identity of Human Environmental Material for Females",
       xlab = "Percent Identity",
       col = "pink")
```


The data next needed to be subsetted by human genomic data which was done though the extrapolation of "Sebum," utilizing the sequence length for each sequence as a secondary check against percent identity matches. These two characters were constantly used to analyze the data throughout the experiment.


**Figure 2a.** Histogram of Query Sequence Length of Human Environmental Material within Whole Data

```{r, echo=FALSE}
# Histogram to visualize only human samples utilizing Human Skin Oil
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  pull(length) %>%
  hist(main = "Sequence Length of Human Environmental Material",
       xlab = "Length in BP",
       col = "darkseagreen2")
```

Once primarily human genomic data was extrapolated, only the female data was further filtered out utilizing the sequence length for each sequence.

**Figure 2b.** Histgram of Query Sequence Length of Human Environmental Material Analyzed for Females

```{r, echo=FALSE}
# Histogram to visualize length of sequences based on human host and sex
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "female") %>%
  pull(length) %>%
  hist(main = "Sequence Length of Human Environmental Material for Females",
       xlab = "Length in BP",
       col = "pink")
```


###Identifying Specific Bacterial Colonies for Females
Once the data had been subsetted by both sequence length and percent identity match, to visualize potential skews and data indescrepancies, a further analyses was done to quantify the amount of specific bacterial colonies seen on female participant samples.


**Table 1** Most Common Bacterial Communities visualized within Female Participants

```{r summary-2, echo=FALSE}

# Table to visualize which Scientific Names are seen within this subset
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum" & sex_s == "female") %>%
group_by(sscinames) %>%
  count() %>%
  arrange(desc(n)) %>%
  head (5) %>%
  kable(col.names = c("Scientific Names", "Count"))

```

**Figure 3** Bar Graph of Most Common Bacterial Communities visualized within Female Participants

```{r, echo=FALSE}
# Bar-chart to visualize most common Scientific Names utilized for female human host participants
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum" & sex_s == "female") %>%
group_by(sscinames) %>%
  count() %>%
  arrange(desc(n)) %>%
  head (5) %>%
  ggplot(aes( x = sscinames,
             y = n)) +
    geom_bar(stat = "Identity", 
             fill = "hotpink") +
    ggtitle("Most Common Bacterial Communities for Females") +
    theme_minimal() + 
    xlab( "Scientific Names") +
    ylab ("Number of Communities") 
```

After the most common bacteiral species was identified for the sex of interest analysis of the data was done through a histogram to visualize the frequency of this specific bacterial community, in regards to percent identity and base pair length within the sequences.

**Figure 4a.** Histogram for Percent Identity of Human Environmental Material Analyzed for Females for Most Common Bacterial Community Found

```{r, echo=FALSE}
# Histogram for most common species visualized by percent identity for females
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "female") %>%
  filter(sscinames == "Bartonella washoensis") %>%
  pull(pident) %>%
  hist(main = "Percent Identity for Females for Most Common Bacteria",
       xlab = "Percent",
       col = "hotpink")
```

**Figure 4b.** Histogram for Query Sequence Length of Human Environmental Material Analyzed for Females for Most Common Community Found

```{r, echo=FALSE}
# Histogram for most common species visualized by length for females
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "female") %>%
  filter(sscinames == "Bartonella washoensis") %>%
  pull(length) %>%
  hist(main = "Sequence Length for Females for Most Common Bacteria",
       xlab = "Length in BP",
       col = "hotpink")
```


######Identifying Specific Bacterial Colonies for Males

After female analyses were run it was concluded that a potential difference could be present between the sexes due to behavioral and physiological traits. To further investigate these differences within sex analyses were continued on male participant data using the most common bacterial species.


**Table 2** Most Common Bacterial Communities visualized within Male Participants

```{r, echo=FALSE}

# Table to visualize which Scientific Names are seen within this subset for males to compare to females
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum" & sex_s == "male") %>%
group_by(sscinames) %>%
  count() %>%
  arrange(desc(n)) %>%
  head (5) %>%
  kable(col.names = c("Scientific Names", "Count"))

```


**Figure 5.** Bar Graph of Most Common Bacterial Communities visualized within Male Participants

```{r, echo=FALSE}

# Bar-chart for most commonly found bacteria for males
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum" & sex_s == "male") %>%
group_by(sscinames) %>%
  count() %>%
  arrange(desc(n)) %>%
  head (5) %>%
  ggplot(aes(x = sscinames,
             y = n)) +
    geom_bar(stat = "Identity",
             fill = "steelblue1") +
    ggtitle("Most Common Bacterial Communities for Males") +
    theme_minimal() + 
    xlab( "Scientific Names") +
    ylab ("Number of Communities")
```


After the most common bacteiral species was identified for the sex of interest analysis of the data was done through a histogram to visualize the frequency of this specific bacterial community, in regards to percent identity and base pair length within the sequences. This was done specifically for the male population due to the visualization of differing bacterial communities from the female specific analyses done prior.


**Figure 6a.** Histogram of Percent Identity of Human Environmental Material Analyzed for Males for Most Common Community Found

```{r, echo=FALSE}

# Histogram for percent identification of specific bacteria found for males
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "male") %>%
  filter(sscinames == "Solemya pervernicosa gill symbiont") %>%
  pull(pident) %>%
  hist(main = "Percent Identity for Males for Most Common Bacteria",
       xlab = "Percent",
       col = "steelblue1")
```


**Figure 6b.** Histogram of Query Sequence Length of Human Environmental Material Analyzed for Males for Most Common Community Found

```{r, echo=FALSE}

# Histogram for seqeunce length of specific bacteria found for males
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "male") %>%
  filter(sscinames == "Solemya pervernicosa gill symbiont") %>%
  pull(length) %>%
  hist(main = "Sequence Length for Males for Most Common Bacteria",
       xlab = "Length in BP",
       col = "steelblue1")
```


After both male and female data sets were individually analyzed comparitive measures were taken to showcase differences in percent identity and seqeunce lengths overall. In this way sex specific interpretations could be made towards bacterial communities visualized based on original sequence data received.

**Figure 7a.** Percent Match for Male Verus Female Participants

```{r, echo=FALSE}

# Boxplot for male versus female percent identification 
ggplot(joined_blast_data_metadata,
       aes(x = sex_s,
           y = pident)) + 
  geom_jitter(alpha = 0.05) +
  ggtitle("Percent Match for Male Verus Female Participants") +
  xlab("Sex") +
  ylab("Percent Match") +
  theme_minimal()
```

**Figure 7b.** Length of Sequences for Male versus Female Participants

```{r, gplot2, echo=FALSE}

# Boxplot for male versus female sequence lengths
ggplot(joined_blast_data_metadata,
       aes(x = sex_s,
           y = length)) + 
  geom_jitter(alpha = 0.05) +
  ggtitle("Length of Sequences for Male versus Female Participants") +
  xlab("Sex") +
  ylab("Length of Sequences") +
  theme_minimal()
```


# Discussion

Add 2-3 paragraphs here interpreting your results and considering future directions one might take in analyzing these data.

