---
title: "Analysis of BLAST Results"
author: "Katelyn B. Kukar"
date: "October 6, 2017"
output: github_document
---

# Introduction


Creating alternative ways of identification is highly important within the field of Forensics because this would allow for more specific analysis of potential suspects. By being able to compare blood, fingerprints, and other means of testing methods allows for a higher level of statistical significance for true suspect identification. Creating alternative methods is also important since some investigations of suspects of interest lack both DNA evidence and individual fingerprints allowing for ambiguity that could be resolved via other testing means. In Fierer et al., scientists explain that bacteria colonized on individual hands represent highly specific and diverse communities. They believe that "because skin bacterial communities are personalized, [they] could use the residual skin bacteria left on objects for forensic identification, matching the bacteria on the object to the skin-associated bacteria of the individual who touched the object," (Fierer et al.). 

Within this study we analyze data collected from the Fierer Lab in Colorado through sequencing and computational techniques. The information gathered from the sequencing sets will be evaluated to further evidence towards the validity of utilizing such methods within forensic analysis. Specifically this analysis will utilize prior knowledge geared towards differences between the male and female sex. We hypothesize the female sex will have significantly different data than that of the male sex due to physiological and behavioral differences. Certain assumptions for this hypothesis include males having traditionally different hand shape and size, body hair, and fat percentage which constitute physiological differences. Assumptions of behavior for these differences are based on hygiene patterns experienced by males and females due to cultural biases towards the feminine construct.


# Methods

Sample origin, sequencing and computational techniques were utilized to analyze sequencing sets collected from human participants and their coinciding bacterial sequences from taken samples.


## Sample origin and sequencing

####Sample Collection

All analysis was done using Fierer et al. as a reference source for sequence data and sample origins. Within the paper it is stated that samples were taken from nine healthy adults (four female, five male) who worked within the same building at the University of Colorado between ages 18-40. The samples taken included computer mouse swabs touched by the owner 12h before swabbing and the palm of the individual's dominant hand used to operate the mouse. Individuals were required to maintain normal hygiene habits prior to the swab as to not increase variation. The swabs were taken using autoclave cotton-tipped swabs that were autoclave and pre-moistened with a sterile solution. The swabbing took place for 10s per sample taken with all samples being kept at -80C for less than a week before DNA extraction (Fierer et al). 

####DNA Extraction and Pyrosequencing

Fierer et al. explains that for each sample the 16s rRNA genes were amplified using the MO BIO PowerSoil DNA Isolation kit with the broken, frozen cotton swabs. These tubes were then horizontally shaken and kit procedures were followed for extraction. PCR reactions were carried out in triplicate repeats using HotMAsterMic with thermal cycling at 94C for 3min followed by 35 cycles of denaturation, annealing for 30s, and extension and final extension procedures. Replicate amplicons were pooled in agarose gel using SYBR Safe DNA gel stain from Invitrogen. Amplicon DNA concentrations were then measured with the final pool of DNA being precipitated, centrifuged, and centrifuged to create a pellet that was re-suspended in nuclease-free water. Pyrosequencing was then carried out on a 454 Life Sciences Genome Sequencer FLX instrument at the University of South Carolina (Fierer et al).


## Computational

In order to analyze the sequences received we used computational programs to evaluate variables of interest. The variables we want to analyze more specifically include participant sex and commonly found bacterial community strains within the metadata set. 
  
To create a meta data set with all of the vectors of interest we initially utilized the raw data sets collected from the 454 Sequencer and published by Fierer et al. This was done using a remote server and fastq-dump to download the list of files in the run table to the raw directory. QC reports were then created for each of the runs utilizing fastqc and outputted as HTML to be readable. The fastqc files were trimmed (Trimmomatic Tool) based on their quality scores to discard sequences under 150 base pairs and cut off reads when the base score drops below 25. Fastqc files were then converted to fasta files to be used as BLAST queries (Bioawk). These files were then run against NCBI database (nt), using BLAST (blastn) to output csv file which was based on the name output and included all of the sequence data, plus the bacterial communities each sequence matched to with their quality scores and additional information.
  
Then the Blast results were further formatted to create an output dataframe with proper column headers that denote each variable of interest from the results (qseqid). After this, the query sequence ID was split into Sample and Number components to group the files into columns (tidyr and separate) to make a vector of all the BLAST output file names, including the names of the directories. An empty matrix was created to join the data back and read in the meta data (rbind).

We than use dplyr piping syntax to select a subset of row matching each vector we want to filter (categorically) and pipe out (numerically) to create histograms. We also used ggplot syntax as a comparative feature to plot multiple categorical variables against one another using one primary numerical variable. Tables were created with kable to tabularize quantified data sets and visualize the most common bacterial communities found for singular categorical vectors (males and females).


# Results

The results section was primarily used to extrapolate data in regards to female participants and the specific bacterial communities identified. This section outlines statistical analyses used to filter through the data set to identify the specific bacterial communities and where they arose against male counterparts.

```{r load-libraries, message = FALSE, echo=FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the intall.packages() function
# or with the 'Packages' pane in RStudio

# load packages
library("dplyr")
library("tidyr")
library("knitr")
library("ggplot2")
```

```{r make-read-in-data-function, echo=FALSE}
# Output format from BLAST is as detailed on:
# https://www.ncbi.nlm.nih.gov/books/NBK279675/
# In this case, we used: '10 sscinames std'
# 10 means csv format
# sscinames means unique Subject Scientific Name(s), separated by a ';'
# std means the standard set of result columns, which are:
# 'qseqid sseqid pident length mismatch
# gapopen qstart qend sstart send evalue bitscore',


# this function takes as input a quoted path to a BLAST result file
# and produces as output a dataframe with proper column headers
# and the 'qseqid' column split into sample and seq number
read_blast_output <- function(filename) {
  data_in <- read.csv(filename,
                      header = FALSE, # files don't have column names in them
                      col.names = c("sscinames", # unique Subject Sci Name(s)
                                    "qseqid",    # Query Seq-id
                                    "sseqid",    # Subject Seq-id
                                    "pident",    # Percntge of identical matches
                                    "length",    # Alignment length
                                    "mismatch",  # Number of mismatches
                                    "gapopen",   # Number of gap openings
                                    "qstart",    # Start of alignment in query
                                    "qend",      # End of alignment in query
                                    "sstart",    # Start of alignment in subj
                                    "send",      # End of alignment in subject
                                    "evalue",    # Expect value
                                    "bitscore"))  # Bit score

  # Next we want to split the query sequence ID into
  # Sample and Number components so we can group by sample
  # They originally look like "ERR1942280.1"
  # and we want to split that into two columns: "ERR1942280" and "1"
  # we can use the separate() function from the tidyr library to do this
  # Note that we have to double escape the period for this to work
  # the syntax is
  # separate(column_to_separate,
  # c("New_column_name_1", "New_column_name_2"),
  # "seperator")
  data_in <- data_in %>%
    separate(qseqid, c("sample_name", "sample_number"), "\\.")
}
```

```{r read-in-BLAST-data,echo=FALSE}
# this makes a vector of all the BLAST output file names, including
# the name(s) of the directories they are in
files_to_read_in <- list.files(path = "output/blast",
                               full.names = TRUE)

# We need to create an empty matrix with the right number of columns
# so that we can rbind() each dataset on to it
joined_blast_data <- matrix(nrow = 0,
                            ncol = 14)

# now we loop over each of the files in the list and append them
# to the bottom of the 'joined_blast_data' object
# we do this with the rbind() function and the function we
# made earlier to read in the files, read_blast_output()
for (filename in files_to_read_in) {
  joined_blast_data <- rbind(joined_blast_data,
                             read_blast_output(filename))
}
```

```{r read-in-metadata-and-join, echo=FALSE}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the `sep = "\t"` tells the function that the data are tab-delimited
# and the `stringsAsFactors = FALSE` tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                                 "fierer_forensic_hand_mouse_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE)

# Finally we use the left_join() function from dplyr to merge or 'join' the
# combined data and metadata into one big table, so it's easier to work with
# in R the `by = c("Run_s" = "sample_name")` syntax tells R which columns
# to match up when joining the datasets together
joined_blast_data_metadata <- metadata_in %>%
  left_join(joined_blast_data,
            by = c("Run_s" = "sample_name"))
```

###Extrapolating Human Female Data from Meta Data

The data first needed to be filtered by human genomic data which was done though the extrapolation of "Sebum," utilizing the percent identification for each sequence.

**Figure 1a.** Histogram of Percent Identity of Human Environmental Material within Whole Data

```{r histograms, echo=FALSE}
# Here we're using the dply piping syntax to select a subset of rows matching a
# criteria we specify (using the filter) function, and then pull out a column
# from the data to make a histogram. We don't need to tell the hist() function
# which data to use, because that's piped in, but we do have to give the
# hist() function the title and axis label we'd like to use for the figure

# Histogram to visualize only human samples utilizing Human Skin Oil
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  pull(pident) %>%
  hist(main = "Percent Identity of Human Environmental Material",
       xlab = "Percent Identity",
       col = "darkseagreen2")
```

Once primarily human genomic data was extrapolated, only the female data was further filtered out utilizing the percent identification for each sequence

**Figure 1b.** Histogram of Percent Identity of Human Environmental Material Analyzed for Females

```{r, echo=FALSE}
# Histogram to visualize sex of human hosts
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "female") %>%
  pull(pident) %>%
  hist(main = "Percent Identity of Human Environmental Material for Females",
       xlab = "Percent Identity",
       col = "pink")
```


The data next needed to be filtered by human genomic data which was done though the extrapolation of "Sebum," utilizing the sequence length for each sequence as a secondary check against percent identity matches. These two characters were constantly used to analyze the data throughout the experiment.


**Figure 2a.** Histogram of Query Sequence Length of Human Environmental Material within Whole Data

```{r, echo=FALSE}
# Histogram to visualize only human samples utilizing Human Skin Oil
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  pull(length) %>%
  hist(main = "Sequence Length of Human Environmental Material",
       xlab = "Length in BP",
       col = "darkseagreen2")
```

Once primarily human genomic data was extrapolated, only the female data was further filtered out utilizing the sequence length for each sequence.

**Figure 2b.** Histogram of Query Sequence Length of Human Environmental Material Analyzed for Females

```{r, echo=FALSE}
# Histogram to visualize length of sequences based on human host and sex
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "female") %>%
  pull(length) %>%
  hist(main = "Sequence Length of Human Environmental Material for Females",
       xlab = "Length in BP",
       col = "pink")
```


###Identifying Specific Bacterial Colonies for Females
Once the data had been filtered by both sequence length and percent identity match, to visualize potential skews and data discrepancies, a further analyses was done to quantify the amount of specific bacterial colonies seen on female participant samples.


**Table 1** Most Common Bacterial Communities visualized within Female Participants

```{r summary-2, echo=FALSE}

# Table to visualize which Scientific Names are seen within this subset
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum" & sex_s == "female") %>%
group_by(sscinames) %>%
  count() %>%
  arrange(desc(n)) %>%
  head (5) %>%
  kable(col.names = c("Scientific Names", "Count"))

```

**Figure 3** Bar Graph of Most Common Bacterial Communities visualized within Female Participants

```{r, echo=FALSE}
# Bar-chart to visualize most common Scientific Names utilized for female human host participants
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum" & sex_s == "female") %>%
group_by(sscinames) %>%
  count() %>%
  arrange(desc(n)) %>%
  head(5) %>%
  ggplot(aes(x = sscinames,
             y = n)) +
    geom_bar(stat = "Identity",
             fill = "hotpink") +
    ggtitle("Most Common Bacterial Communities for Females") +
    theme_minimal() +
    xlab("Scientific Names") +
    ylab("Number of Communities")
```

After the most common bacterial species was identified for the sex of interest analysis of the data was done through a histogram to visualize the frequency of this specific bacterial community, in regards to percent identity and base pair length within the sequences.

**Figure 4a.** Histogram for Percent Identity of Human Environmental Material Analyzed for Females for Most Common Bacterial Community Found

```{r, echo=FALSE}
# Histogram for most common species visualized by percent identity for females
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "female") %>%
  filter(sscinames == "Bartonella washoensis") %>%
  pull(pident) %>%
  hist(main = "Percent Identity for Females for Most Common Bacteria",
       xlab = "Percent",
       col = "hotpink")
```

**Figure 4b.** Histogram for Query Sequence Length of Human Environmental Material Analyzed for Females for Most Common Community Found

```{r, echo=FALSE}
# Histogram for most common species visualized by length for females
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "female") %>%
  filter(sscinames == "Bartonella washoensis") %>%
  pull(length) %>%
  hist(main = "Sequence Length for Females for Most Common Bacteria",
       xlab = "Length in BP",
       col = "hotpink")
```


######Identifying Specific Bacterial Colonies for Males

After female analyses were run it was concluded that a potential difference could be present between the sexes due to behavioral and physiological traits. To further investigate these differences within sex analyses were continued on male participant data using the most common bacterial species.


**Table 2** Most Common Bacterial Communities visualized within Male Participants

```{r, echo=FALSE}

# Table for males to compare to females for Sci Names
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum" & sex_s == "male") %>%
group_by(sscinames) %>%
  count() %>%
  arrange(desc(n)) %>%
  head(5) %>%
  kable(col.names = c("Scientific Names", "Count"))

```


**Figure 5.** Bar Graph of Most Common Bacterial Communities visualized within Male Participants

```{r, echo=FALSE}

# Bar-chart for most commonly found bacteria for males
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum" & sex_s == "male") %>%
group_by(sscinames) %>%
  count() %>%
  arrange(desc(n)) %>%
  head(5) %>%
  ggplot(aes(x = sscinames,
             y = n)) +
    geom_bar(stat = "Identity",
             fill = "steelblue1") +
    ggtitle("Most Common Bacterial Communities for Males") +
    theme_minimal() +
    xlab("Scientific Names") +
    ylab("Number of Communities")
```


After the most common bacterial species was identified for the sex of interest analysis of the data was done through a histogram to visualize the frequency of this specific bacterial community, in regards to percent identity and base pair length within the sequences. This was done specifically for the male population due to the visualization of differing bacterial communities from the female specific analyses done prior.


**Figure 6a.** Histogram of Percent Identity of Human Environmental Material Analyzed for Males for Most Common Community Found

```{r, echo=FALSE}

# Histogram for percent identification of specific bacteria found for males
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "male") %>%
  filter(sscinames == "Solemya pervernicosa gill symbiont") %>%
  pull(pident) %>%
  hist(main = "Percent Identity for Males for Most Common Bacteria",
       xlab = "Percent",
       col = "steelblue1")
```


**Figure 6b.** Histogram of Query Sequence Length of Human Environmental Material Analyzed for Males for Most Common Community Found

```{r, echo=FALSE}

# Histogram for seqeunce length of specific bacteria found for males
joined_blast_data_metadata %>%
  filter(env_material_s == "sebum") %>%
  filter(sex_s == "male") %>%
  filter(sscinames == "Solemya pervernicosa gill symbiont") %>%
  pull(length) %>%
  hist(main = "Sequence Length for Males for Most Common Bacteria",
       xlab = "Length in BP",
       col = "steelblue1")
```


After both male and female data sets were individually analyzed comparative measures were taken to showcase differences in percent identity and sequence lengths overall. In this way sex specific interpretations could be made towards bacterial communities visualized based on original sequence data received.

**Figure 7a.** Percent Match for Male Versus Female Participants

```{r, echo=FALSE}

# Boxplot for male versus female percent identification
ggplot(joined_blast_data_metadata,
       aes(x = sex_s,
           y = pident)) +
  geom_jitter(alpha = 0.05) +
  ggtitle("Percent Match for Male Verus Female Participants") +
  xlab("Sex") +
  ylab("Percent Match") +
  theme_minimal()
```

**Figure 7b.** Length of Sequences for Male versus Female Participants

```{r, gplot2, echo=FALSE}

# Boxplot for male versus female sequence lengths
ggplot(joined_blast_data_metadata,
       aes(x = sex_s,
           y = length)) +
  geom_jitter(alpha = 0.05) +
  ggtitle("Length of Sequences for Male versus Female Participants") +
  xlab("Sex") +
  ylab("Length of Sequences") +
  theme_minimal()
```


# Discussion

####Females
  Utilizing R as a tool to create our graphics for statistical analysis of the sequence data we were able to extrapolate multiple histograms that depicted skews related to percent identity and sequence length in female participants. 
    We were able to extrapolate only human samples from the data set by looking for "sebum" samples, these samples represent a common human skin oil which is unlike the dust sampled from the electronics.
    For percent identity of the specific sequences found the histogram represented a majority of the sequences residing in the 98-100 percent identity match range. We can conclude that for Female samples the sequences were more specific. To create a secondary check for the goodness of the sequence data we also looked at sequence length to see if any short sequences potentially created dependencies during blast analysis. The histogram depicted a skew with the majority of sequences between 220-240 base pairs. However marginal outliers did exist. We can rationalize these sequence lengths because a majority of sequencing platforms can only sequence a maximum read length of 240 base pairs. 
    We finally filtered out the female participants in regard to the scientific name of the species and count the most common bacterial communities in order to divulge the most common species found was **Bartonella washoensis**. **Bartonella washonesis** is commonly found as a cause for meningitis. Potentially the female participants were more susceptible to meningitis due to their proximity of work (in the same building) or from exposure to mites from animals located at the University of Colorado, a naturally "outdoors" oriented campus. 
    This specific data was analyzed further to find the average percent match for **Bartonella washoensis** was between 99-100 percent, and sequence length in base pairs was between 220-230. Both of these numbers give us evidence towards the fact that this is an indisputable found.

####Males
  Once female data was compiled male participants were also analyzed to find **Solemya pervernicosa gill symbiont** as the primary bacterial colony for male only samples. This bacterial community resides in the taxa of sulfur-oxidizing symbionts. This means this bacteria usually requires a host environment, where in which, it's cohabitation benefits the host. The sulfur-oxidizing symbionts tend to inhabit marine hosts, much as the classification **Soleyma pervernicosa** defines. **Soleyma pervernicosa** is a small bivalve, salt-water clam that resides in shallow marine habitats. Concluding that the sulfur-oxidizing feature of this symbiont is to give nutrition to the bivalve since it naturally lacks a digestive system. 
    Then when the most common bacterial community was graphed as a histogram it was found the percent identity was only between 86-87 percent, very unlike the percent of the female. The histogram for sequence was only found to be under 160 base pair lengths.
    This information is incredibly concerning since both the percent identity and base pair lengths are far lower. We can conclude this sequence data could be inconclusive because of too little DNA, an unable to be sequenced region, a "dirty" template, sample contamination, or other issues. This should be re-sequenced or lab notebook notes should be reevaluated for error.

